src/shogun/mathematics/linalg/internal/implementation/Mean.h

```diff
@@ -0,0 +1,984 @@
+ * as backend for computing mean.
+ */
+template <class Matrix>
+struct mean_symmetric<Backend::EIGEN3,Matrix>
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/**
+	 * Method that computes the mean of symmetric SGMatrix using Eigen3
+	 *
+	 * @param m the matrix whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}m_{i,j}\f$
+	 */
+	static T compute(SGMatrix<T> mat, bool no_diag)
+	{
+                REQUIRE(mat.num_rows > 0, "Matrix can not be empty!\n");
+                
+                // remove the main diagonal elements if required
+                if (no_diag) 
+                {
+                    return (sum_symmetric<Backend::EIGEN3, SGMatrix<T> >
+                    ::compute(mat, no_diag) / (mat.num_rows * (mat.num_cols - 1)));
+                }
+                else
+                {
+                    return (sum_symmetric<Backend::EIGEN3, SGMatrix<T> >
+                    ::compute(mat, no_diag) / (mat.num_rows * mat.num_cols));
+                }
+	}
+
+	/**
+	 * Method that computes the mean of symmetric SGMatrix blocks using Eigen3
+	 *
+	 * @param b the matrix-block whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}b_{i,j}\f$
+	 */
+	static T compute(Block<SGMatrix<T> > b, bool no_diag)
+	{
+		REQUIRE(b.m_row_size > 0, "Matrix can not be empty!\n");
+                
+                // remove the main diagonal elements if required
+                if (no_diag) 
+                {
+                    return (sum_symmetric<Backend::EIGEN3, Block<SGMatrix<T> > >
+                    ::compute(b, no_diag) / (b.m_row_size * (b.m_col_size - 1)));
+                }
+                else
+                {
+                    return (sum_symmetric<Backend::EIGEN3, Block<SGMatrix<T> > >
+                    ::compute(b, no_diag) / (b.m_row_size * b.m_col_size));
+                }
+	}
+};
+
+/**
+ * @brief Specialization of generic colwise_mean which works with SGMatrix and uses Eigen3
+ * as backend for computing mean.
+ */
+template <class Matrix>
+struct colwise_mean<Backend::EIGEN3,Matrix> ////////////////// SGVector rowwise divide
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/** Return type */
+	typedef SGVector<T> ReturnType;
+
+	/** Eigen vector type */
+	typedef Eigen::Matrix<T,Eigen::Dynamic,1> VectorXt;
+
+	/**
+	 * Method that computes the column wise mean of SGMatrix using Eigen3
+	 *
+	 * @param m the matrix whose colwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the colwise mean computed as \f$s_j=\mean_{i}m_{i,j}\f$
+	 */
+	static SGVector<T> compute(SGMatrix<T> m, bool no_diag)
+	{
+		REQUIRE(m.num_cols > 0, "Matrix can not be empty!\n");
```

Comment:
Can you state that you checked the number of columns here? Also above
"Number of matrix columns (%d) must not be 0" % m.num_cols